### PR TITLE
feat(lexer): support new RegExp Indices flag "d"

### PR DESCRIPTION
**Pull request by 3cp**
Related to #214

### CODE REVIEW

**Comment by 3cp** [comment]
Not behind `next` since it's a stage 4 feature.

File: test/lexer/regexp.ts [hunk 1]
```diff
@@ -171,21 +173,19 @@ describe('Lexer - Regular expressions', () => {
       t.throws(() => scanSingleToken(state, context, 0));
     });
   }
-  fail('fails on /i/ii', '/ ', Context.AllowRegExp);
```
Comment (3cp) [inline test/lexer/regexp.ts:174]: There are lots of wrong message in this sections. And there are many duplicated tests on `/i/yy` and `/i/ss`.

**Review by aladdin-add — APPROVED** [review]

File: src/lexer/regexp.ts [hunk 2]
```diff
@@ -136,6 +142,8 @@ export function scanRegularExpression(parser: ParserState, context: Context): To
  */
 function validate(parser: ParserState, pattern: string, flags: string): RegExp | null | Token {
   try {
+    // Temporarily allows older version of Nodejs (or browser) to accept the new Indices flag "d".
+    flags = flags.replace('d', '');
```
Comment (aladdin-add) [inline src/lexer/regexp.ts:146]: will it return a wrong value, even if the `/d` is supported?

Comment (3cp) [reply src/lexer/regexp.ts:146]: Pls elaborate.

Comment (aladdin-add) [reply src/lexer/regexp.ts:146]: e.g. `value: /./d` =>  `/./` (as the `'/d'` has been replaced)?

Comment (aladdin-add) [reply src/lexer/regexp.ts:146]: ok. seems it won't happen (as stated in the valid test). :)

**Review by aladdin-add — APPROVED** [review]

**Comment by 3cp** [comment]
I overlooked the validation result is used.
```js
parser.tokenValue = validate(parser, pattern, flags);
```
So the tokenValue is wrong for regex with 'd' flag. Not sure what's the best solution. It might need a try catch to fall back to removing "d", so that at least for Nodejs v18, the tokenValue is correct.

**Comment by aladdin-add** [comment]
a straightforward way is to do a check in the entry, something like:

```js
var isDsupported = true;
try {new Regexp('.', 'd');}catch(e){isDsupported = false;}
```

we may also want to check others like `/s`, `/u`, `/y`...

btw, I don't think it's a good resolution to just replace them to '', as they may be misused as normal values. e.g. `/./d` => `/./`. what do you think just return an exceptional value, e.g. `null`?

**Comment by fisker** [comment]
Is this stored as RegExpLiteral value? It should be null per ESTree spec https://github.com/estree/estree/blob/master/es5.md#regexpliteral

**Comment by 3cp** [comment]
```
> r.body[0].expression
{ type: 'Literal', value: /a/g, regex: { pattern: 'a', flags: 'g' } }
```

The value is for parent interface just above your link.

```
interface Literal <: Expression {
    type: "Literal";
    value: string | boolean | null | number | RegExp;
}
```

**Comment by 3cp** [comment]
Oh, sorry. I got it now.

**Comment by 3cp** [comment]
What about
```js
try {
  return new RegExp(pattern, flags);
} catch (e) {
  try {
    // Fallback to only validate patterns on systems 
    // where some flags are not supported.
    new RegExp(pattern);
    return null;
  } catch (e) {
    report(parser, Errors.UnterminatedRegExp);
  }
}
```

**Comment by aladdin-add** [comment]
it will allow unknown/dulp flags, e.g. `/foo/a`, `/foo/gg`?

**Comment by 3cp** [comment]
Validation happens after the duplication check.

**Pull request merged**